cpp/src/arrow/util/bit_stream_utils.h

```diff
@@ -341,11 +344,27 @@ inline int BitReader::GetBatch(int num_bits, T* v, int batch_size) {
   }
 
   if (sizeof(T) == 4) {
-    int num_unpacked =
-        internal::unpack32(reinterpret_cast<const uint32_t*>(buffer + byte_offset),
-                           reinterpret_cast<uint32_t*>(v + i), batch_size - i, num_bits);
-    i += num_unpacked;
-    byte_offset += num_unpacked * num_bits / 8;
+    if (num_bits <= 16) {
+#if defined(ARROW_HAVE_AVX512_ICX)
```

Comment:
I am afraid that more if/else will be added here to support more instruction sets, which makes it less maintainable. Is it possible to move the dispatch logic into `internal::unpack32` implementation and let the dispatcher do the job? In this way, we can avoid exposing `arrow/util/bpacking_avx512_icx.h` in this public header as well.

Reply:
@wgtmac I agree. It is better to put the logic into the dispatcher instead of the bit_stream_utils.h file. Code has been updated. Please review it. Thanks.